examples/iso_dh/DY.Example.DH.Protocol.Total.fst

```diff
@@ -0,0 +1,167 @@
+[@@ with_bytes bytes]
+type message1 = {
+  alice:principal;
+  gx:bytes;
+}
+
+%splice [ps_message1] (gen_parser (`message1))
+%splice [ps_message1_is_well_formed] (gen_is_well_formed_lemma (`message1))
+
+[@@ with_bytes bytes]
+type message2 = {
+  bob:principal;
+  gy:bytes;
+  sg:bytes;
+}
+
+%splice [ps_message2] (gen_parser (`message2))
+%splice [ps_message2_is_well_formed] (gen_is_well_formed_lemma (`message2))
+
+[@@ with_bytes bytes]
+type message3 = {
+  sg:bytes;
+}
+
+%splice [ps_message3] (gen_parser (`message3))
+%splice [ps_message3_is_well_formed] (gen_is_well_formed_lemma (`message3))
+
+
+[@@ with_bytes bytes]
+type message =
+  | Msg1: msg:message1 -> message
+  | Msg2: msg:message2 -> message
+  | Msg3: msg:message3 -> message
+
+%splice [ps_message] (gen_parser (`message))
+%splice [ps_message_is_well_formed] (gen_is_well_formed_lemma (`message))
+
+instance parseable_serializeable_message: parseable_serializeable bytes message = mk_parseable_serializeable ps_message
+
+// Definition of signature terms
+[@@ with_bytes bytes]
+type sig_message2 = {
+  alice:principal;
+  gx:bytes;
+  gy:bytes;
+}
+
+%splice [ps_sig_message2] (gen_parser (`sig_message2))
+%splice [ps_sig_message2_is_well_formed] (gen_is_well_formed_lemma (`sig_message2))
+
+[@@ with_bytes bytes]
+type sig_message3 = {
+  bob:principal;
+  gx:bytes;
+  gy:bytes;
+}
+
+%splice [ps_sig_message3] (gen_parser (`sig_message3))
+%splice [ps_sig_message3_is_well_formed] (gen_is_well_formed_lemma (`sig_message3))
+
+[@@ with_bytes bytes]
+type sig_message =
+  | SigMsg2: msg:sig_message2 -> sig_message
+  | SigMsg3: msg:sig_message3 -> sig_message
+
+%splice [ps_sig_message] (gen_parser (`sig_message))
+%splice [ps_sig_message_is_well_formed] (gen_is_well_formed_lemma (`sig_message))
+
+instance parseable_serializeable_sig_message: parseable_serializeable bytes sig_message = mk_parseable_serializeable ps_sig_message
+
+
+(*** Message Processing ***)
+
+// Alice generates message 1
+val compute_message1: principal -> bytes -> bytes
+let compute_message1 alice x =
+  let gx = dh_pk x in
+  let msg = Msg1 {alice; gx} in
+  serialize message msg
+
+// Bob parses message 1
+val decode_message1: bytes -> option message1
+let decode_message1 msg1_bytes =
+  let? msg1 = parse message msg1_bytes in
+  // These lines are the...
+  guard (Msg1? msg1);?
+  Some (Msg1?.msg msg1)
+  // ...short version of the following match:
+  (*    
+  match msg1 with
+  | Msg1 msg -> Some msg
+  | _ -> None
+  *)
+
+// Bob generates message 2
+val compute_message2: principal -> principal -> bytes -> bytes -> bytes -> bytes -> bytes
+let compute_message2 alice bob gx gy sk_b n_sig =
+  let sig_msg = SigMsg2 {alice; gx; gy} in
+  let sg = sign sk_b n_sig (serialize sig_message sig_msg) in
+  let msg = Msg2 {bob; gy; sg} in
+  serialize message msg
+
+// Alice parses message 2
+type verify_msg2_result = {sg:bytes; gy:bytes; gx:bytes; k:bytes}
```

Comment:
I feel like the function should not return the signature, as it is something internal to the protocol.
`gx` and `gy` make sense since it corresponds to the identity of Alice and Bob, and also `k` because it is the shared secret.

Reply:
I can see this, but leaving the signature out makes the proof code more complex because I have to parse the message manually in the proof ``decode_and_verify_message2_proof``.